### ISSUE TITLE
TxId omitido nos comprovantes e extratos

### ISSUE THREAD

**Issue by flaviolenz** [bacen/pix-api#184]
Pelo menos pra Pessoa Fisica, pelo menos 3 bancos que testei estao omitindo o TxId dos comprovantes e extratos.
A transação eh iniciada por QR estático.

Pode isso?

**Comment by renatofrota** [comment]
Também testei algumas coisas e já encontrei:

* chaves sendo registradas somente com a razão social (sendo que a empresa tem nome fantasia)
* txid sendo ignorado totalmente na leitura do QR Code estático (campo "mensagem" ficando vazio e editável), o que é pior que a "não exibição" no comprovante (pois inviabiliza totalmente a conciliação pelo recebedor)
* opção de leitura de QR Code não disponível dentro da seção Pix (somente por outro caminho do app, totalmente não relacionado com o Pix)
* falta da opção de enviar Pix usando "Pix Copia e Cola"
* impossibilidade de enviar Pix (instituição exigindo que eu cadastre uma chave, o que só deveria acontecer se eu estivesse querendo receber um Pix)

entre outras...

Uma instituição chamou o Copia e Cola de "Pix via código". Pode parecer estranho lendo aqui, mas não se confundiu com a opção de envio por chave e ficou bastante intuitivo na interface, apesar de essa variação de nomenclatura não ser prevista/permitida pelo manual de uso de marca, que eu me lembre (depois vou ler o manual de novo e atualizo esse comentário).

**Comment by rubenskuhl** [comment]
Esse problema do nome fantasia eu tive também, está com o banco e se não resolver, RDR. 
Eu tive limite de envio de envio de Pix em outro (só queria mandar Pix acima de R$0.99), após contato em rede social, informaram que vão corrigir. 

Acho que para tudo que você reportou o roteiro é o mesmo: informa a instituição, se ela não corrigir, abre RDR.

**Comment by renatofrota** [comment]
Encontrei mais problemas:

- bancos considerando dados do QR Code (BR Code) que deveriam ser ignorados no contexto do Pix (ex: o campo Merchant Name, ID 59) em vez de ler as informações do DICT;
- bancos exibindo informações de agência e conta (origem e destino) durante todo o processo de envio e nos comprovantes (em contrariedade ao manual de UX e, inclusive, **ferindo a LGPD**, pois a informação não é essencial para identificar o cliente destinatário e, no momento do cadastramento da chave, é informado que apenas nome completo + CPF parcial + nome da instituição financeira serão exibidos ao pagador);
- bancos omitindo as informações txid e mensagem do campo de texto livre nos comprovantes

Os valores txid e a mensagem do campo de texto livre também estão sendo omitidos no extrato mas o manual de UX, estranhamente, não colocou essa obrigatoriedade. Como se espera que as pessoas identifiquem a que se refere um pagamento sem essas informações?

**Comment by flaviolenz** [comment]
Tem banco não oferecendo a opção de chave aleatória.

100% dos 7 bancos que testei estão omitindo o Txid. Em PJ também.

De que vale ter 200 linhas de extrato com "Recebimento Pix" ?

A especificação deveria determinar que: tendo TXID, a informação deveria
vir obrigatoriamente de maneira estruturada (no caso de API) e textual em
TODAS as notificações e extratos. (Não só no detalhamento)

Saber o nome do estabelecimento a colocar no QR Estático também esta
confuso. Tenho que gerar um QRCode pelo App do banco pra saber o que
colocar aí. Alguns escolhem Nome de Fantasia e outros a Razão Social.
Muitas vezes, com a cadeia quebrada.

Duas duvidas:

1. Podem ir caracteres acentuados no QR? (Cidade, nome e txid)

2. São permitidos espaços, traços e pontuação no TXID?

Em ter, 17 de nov de 2020 06:59, renatofrota <notifications@github.com>
escreveu:

**Comment by renatofrota** [comment]
> Saber o nome do estabelecimento a colocar no QR Estático também esta confuso.

Indiferente. Os dados da chave presente no QR Code devem ser recuperados pelo PSP pagador diretamente do DICT.

> Duas duvidas: 1. Podem ir caracteres acentuados no QR? (Cidade, nome e txid)

Cidade e nome, somente alfanuméricos e espaços.

> 2. São permitidos espaços, traços e pontuação no TXID?

No caso do txid, somente [a-zA-Z0-9] (alfanuméricos sem espaços, com distinção de caixa baixa/alta para as letras).

**Comment by flaviolenz** [comment]
Saber o nome do estabelecimento a colocar no QR Estático também esta
> confuso.
>
> Indiferente. Os dados da chave presente no QR Code devem ser recuperados
> pelo PSP pagador diretamente do DICT.
>

Se for o nome errado, tem banco recusando o QR.

Duas duvidas: 1. Podem ir caracteres acentuados no QR? (Cidade, nome e txid)
>
> Cidade e nome, somente alfanuméricos e espaços.
>

'á' eh alfanumérico?


>    1. São permitidos espaços, traços e pontuação no TXID?
>
> No caso do txid, somente [a-zA-Z0-9] (alfanuméricos sem espaços, com
> distinção de caixa baixa/alta para as letras).
>

Fiz um com espaço no TXID e funcionou.
So nao sei se chegou do outro lado, por causa da omissão,  mas imagino que
sim.

**Comment by renatofrota** [comment]
> Se for o nome errado, tem banco recusando o QR.

Pode passar o código pra gente ver? Se quiser mandar por e-mail, meu usuario @ gmail ponto com.

> Fiz um com espaço no TXID e funcionou.
> So nao sei se chegou do outro lado, por causa da omissão,  mas imagino que sim.

A transmissão de informações em inconformidade é rejeitada (pelo menos deveria ser). Chegou sem o txid no recebedor, provavelmente. E se no momento da leitura o txid foi exibido no app do PSP pagador, isso é um erro de comportamento. Deveria rejeitar o QR (preferencialmente) ou não acatar o txid em inconformidade.

**Comment by flaviolenz** [comment]
Nao consegui repetir o problema. Acho que consertaram.
Gerei um QR com um nome nada a ver e funcionou.
Banco pagador pegou do DICT.
O registro no DICT eh que está sendo feito por alguns bancos como razao social e outros como nome de fantasia.
Acho que poderiam perguntar qual desses o usuario quer, mas nao deixar livre, pra evitar fraudes.

**Comment by renatofrota** [comment]
> Nao consegui repetir o problema. Acho que consertaram.
> Gerei um QR com um nome nada a ver e funcionou.
> Banco pagador pegou do DICT.
> O registro no DICT eh que está sendo feito por alguns bancos como razao social e outros como nome de fantasia.
> Acho que poderiam perguntar qual desses o usuario quer, mas nao deixar livre, pra evitar fraudes.

A regulamentação é usar o Nome Fantasia do CNPJ, quando houver.

**Comment by rubenskuhl** [comment]
> > Nao consegui repetir o problema. Acho que consertaram.
> > Gerei um QR com um nome nada a ver e funcionou.
> > Banco pagador pegou do DICT.
> > O registro no DICT eh que está sendo feito por alguns bancos como razao social e outros como nome de fantasia.
> > Acho que poderiam perguntar qual desses o usuario quer, mas nao deixar livre, pra evitar fraudes.
> 
> A regulamentação é usar o Nome Fantasia do CNPJ, quando houver.

Exato. Mas o Santander não leu o regulamento, e não aceita que esteja errado. Tive que abrir RDR.

**Comment by dev-gto** [comment]
A documentação do ManualBRCode v2.0.1 (em particular na tabela 1, página 10) omitiu os tipos dos campos definidos na norma EMV.

https://www.bcb.gov.br/content/estabilidadefinanceira/spb_docs/ManualBRCode.pdf

Minha sugestão é seguir o que está na norma EMV.

Por enquanto, vi que cidade (tag '60') e merchant Name (tag '59') são do tipo ans (Alphanumeric Special), que não suporta acentos.

Há campos do tipo S (String), e estas suportam unicode (utf-8).

https://www.emvco.com/wp-content/plugins/pmpro-customizations/oy-getfile.php?u=/wp-content/uploads/documents/EMVCo-Merchant-Presented-QR-Specification-v1-1.pdf

No documento acima diz que ans usa o anexo B em:
https://www.emvco.com/wp-content/plugins/pmpro-customizations/oy-getfile.php?u=/wp-content/uploads/documents/EMV_v4.3_Book_4_Other_Interfaces_20120607062305603.pdf

**Comment by dmkamers** [comment]
@dev-gto cabe esclarecer que o padrao BRCode prevalece sempre. 
De fato, a _omissao_ significa que vale o padrao EMV (campos ANS estao bem especificados na referencia que voce apontou).
Sao 96 caracteres que correspondem ao padrao ASCII (7 bits) sem os caracteres de controle (sem acentuacao, portanto).
Ainda que o QRCode EMV utilize UTF-8, campos do tipo `ANS` estao restritos a este conjunto de 96 caracteres.

Sobre apresentar txid ou mesmo descricao para o recebedor, eh importante que os bancos evitem que os pagadores coloquem sequencias perigosas nesses campos, como por exemplo `http://etc` (nao seria apropriado receber um SMS `confiavel` com uma string arbitraria sem nenhum tipo de validacao).

**Comment by rubenskuhl** [comment]
@dmkamers O TxID do QR-Code estático tem restrição de set caracteres mais restritivas que ASCII 7 bits ?

**Comment by renatofrota** [comment]
> @dmkamers O TxID do QR-Code estático tem restrição de set caracteres mais restritiva que ASCII 7 bits ?

Tem. A restrição txid (no Pix em geral, não apenas no QR Code) é `[a-zA-Z0-9]` e a tabela ASCII 7 bits tem [esse e] vários outros caracteres.

**Comment by dmkamers** [comment]
@rubenskuhl excelente pergunta... vou pesquisar antes de responder... mas, a principio, me parece que o caso estático esta menos restritivo que no caso dinamico.

**Comment by dmkamers** [comment]
@rubenskuhl o txid esta limitado no Manual de Padroes para Iniciacao de Pagamento:

> 1.5.2. Identificador de transação: txId no QR Code estático
> O objeto primitivo EMV 62-05 Reference Label, conforme especificado no manual do BR Code, é
> limitado a 25 caracteres e, quando em efeito23, deve ser utilizado para conciliar pagamentos. Trata-se
> de um identificador de transação que deve ser retransmitido intacto pelo PSP do pagador ao gerar a
> ordem de pagamento. Essa informação permitirá ao recebedor identificar e correlacionar a
> transferência, quando recebida, com a apresentação das instruções ao pagador.
> Os caracteres permitidos **no contexto do Pix** para o campo txId (EMV 62-05) são:
> • Letras minúsculas, de ‘a’ a ‘z’
> • Letras maiúsculas, de ‘A’ a ‘Z’
> • Dígitos decimais, de ‘0’ a ‘9’

Obrigado @renatofrota. 
Alias, estou com a impressao que tem txid errado 'a beca por ai quanto a isto... :(

Aproveito para enfatizar mais um conceito confuso @dev-gto : o conteudo de um EMV-QRCPS esta definido como Unicode UTF-8. Ser `alfanumeric-special` *nao* significa que nao poderia ocorrer algum `BOM` por exemplo nos campos ANS. Acredito que a melhor estrategia eh tratar todos os campos como UTF-8 e entao validar (restringir) o conjunto possivel.

**Comment by dev-gto** [comment]
@dmkamers: verdade, o grupo de caracteres ficou ainda mais restrito no BR Code #197

**Comment by renatofrota** [comment]
> @dmkamers: verdade, o grupo de caracteres ficou ainda mais restrito no BR Code #197
> 
> Nomes Fantasia com `&` precisarão de tratamento, creio

Na realidade, no contexto do Pix, o nome **sempre** é obtido do DICT (quando o PSP não comete deslizes. e.g.: Banco Inter). Pode-se utilizar somente "Pix" no QR Code, apenas para cumprimento da necessidade de presença do campo para validação do QR no contexto multi-arranjos e da documentação BR Code, sem prejuízo de conformidade com a regulamentação do Pix.

**Comment by renatofrota** [comment]
> @dev-gto cabe esclarecer que o padrao BRCode prevalece sempre.
> De fato, a _omissao_ significa que vale o padrao EMV (campos ANS estao bem especificados na referencia que voce apontou).
> Sao 96 caracteres que correspondem ao padrao ASCII (7 bits) sem os caracteres de controle (sem acentuacao, portanto).
> Ainda que o QRCode EMV utilize UTF-8, campos do tipo `ANS` estao restritos a este conjunto de 96 caracteres.
> 
> Sobre apresentar txid ou mesmo descricao para o recebedor, eh importante que os bancos evitem que os pagadores coloquem sequencias perigosas nesses campos, como por exemplo `http://etc` (nao seria apropriado receber um SMS `confiavel` com uma string arbitraria sem nenhum tipo de validacao).

Por falar em prevalência do BRCode, poderíamos nos desamarrar do EMV de uma vez... :grin:

Atualizar o BR Code com as mudanças:

* Deixar os campos (que sejam de texto) livres de amarras em relação ao conjunto de caracteres aceitos
* Retirar a obrigatoriedade de campos que não sejam essenciais para cada arranjo em particular (por que exigir a presença do "59040000", um campo que não usamos nunca? ou mesmo o campo 60?)

A interoperabilidade do UTF-8 em todos os tipos de dispositivos comuns atualmente permitiria abolir as limitações de caracteres (e o leitor já precisa suportar UTF-8 para poder decodifir a string inteira, então ter uma limitação de conjunto de caracteres diferente em cada campo já não faz qualquer sentido no próprio EMV).

Ou, seguindo a linha oposta, o campo 59 poderia ter sido utilizado para armazenar o prefixo CNAE do recebedor, quando pessoa jurídica (poderia até ser específico do produto/serviço adquirido em particular, quando atuante com vários CNAE). Seria uma informação muito relevante que hoje o mercado de adquirência tem (e permite categorizar as compras na fatura do cartão de crédito e no extrato bancário).

**Comment by dmkamers** [comment]
@renatofrota qualquer padrão (mesmo ruim) é nosso amigo (vide `problemas de interoperabilidade` com CRC-CCITT).
Possível que existam dispositivos que nao consigam tratar utf8, a ainda assim ler um qr sem engasgar (nao impede CRC correto por exemplo), uma vez que use apenas campos emv numérico e `ans` no arranjo em questão, e ignore strings.
Enfim, creio que o padrão emv resolve um problemao, ao custo de praticamente nao criar outros muito significativos.
Tem NFC vindo ai...

**Comment by alanpcs** [comment]
> Também testei algumas coisas e já encontrei:
> 
> * chaves sendo registradas somente com a razão social (sendo que a empresa tem nome fantasia)
> * txid sendo ignorado totalmente na leitura do QR Code estático (campo "mensagem" ficando vazio e editável), o que é pior que a "não exibição" no comprovante (pois inviabiliza totalmente a conciliação pelo recebedor)
> * opção de leitura de QR Code não disponível dentro da seção Pix (somente por outro caminho do app, totalmente não relacionado com o Pix)
> * falta da opção de enviar Pix usando "Pix Copia e Cola" [edit: o manual de UX não determina obrigatoriedade dessa funcionalidade]
> * impossibilidade de enviar Pix (instituição exigindo que eu cadastre uma chave, o que só deveria acontecer se eu estivesse querendo **receber** um Pix)
> 
> entre outras...
> 
> Uma instituição chamou o Copia e Cola de "Pix via código". Pode parecer estranho lendo aqui, mas não se confundiu com a opção de envio por chave e ficou bastante intuitivo na interface, apesar de essa variação de nomenclatura não ser prevista/permitida pelo manual de uso de marca, que eu me lembre (depois vou ler o manual de novo e atualizo esse comentário).
> 
> -edit- realmente é considerado obrigatório o uso do termo específico "Pix Copia e Cola" (pág 7 do manual de experiência do usuário).

@renatofrota conseguiria o link do manual de experiência do usuário, por gentileza? Pesquisei aqui mas não encontrei algo oficial exatamente assim. Obrigado.

**Comment by renatofrota** [comment]
> @renatofrota conseguiria o link do manual de experiência do usuário, por gentileza? Pesquisei aqui mas não encontrei algo oficial exatamente assim. Obrigado.

Clique [aqui](https://www.bcb.gov.br/estabilidadefinanceira/pix), depois no botão "Regulamentação relacionada ao Pix" (à direita) e terá acesso a todos os documentos vigentes.
